core-rust/state-manager/src/store/historical_state.rs

```diff
@@ -77,48 +79,74 @@ use crate::{
 /// This database is backed by:
 /// - a [`ReadableTreeStore`] - a versioned source of ReNodes / Partitions / Substates metadata,
 /// - and a [`LeafSubstateValueStore`] - a store of Substate values' associated with their leafs.
-pub struct StateTreeBasedSubstateDatabase<'t, T> {
-    tree_store: &'t T,
+pub struct StateTreeBasedSubstateDatabase<'s, DS> {
+    base_store: DS,
     at_state_version: StateVersion,
+
+    // Note: "Why do we even need to capture `'t`?"
```

Comment:
Well, this is because of the fact that e.g. the `SubstateTier`'s `into_iter_substate_summaries_from` method exposes an iterator with a lifetime of the underlying _store_ rather than the entity tier (in other words, it can outlive the tier).

However, the underlying store comes from `self.base_store.deref()` and we need someone to capture the constraint that the reference created lives longer than the `&self` parameter in order to satisfy the `SubstateDatabase` trait definition for `list_entries_from`.

But `self.base_store.deref()` could either resolve to some random lifetime if `base_store` is a reference, or could resolve to `&self.base_store`, so could be bounded by `&self`.

We need some way of capturing that "the lifetime of `self.base_store.deref()`" outlives _ANY_ reference to self. This is actually extra information -- and this phantom lifetime is the best way of doing that.

Reply:
Perhaps it'd be clearer to actually make it a `PhantomData<&'s S>` and have  `'s, S: 's, DS: Deref<Target = S>`.

Or just `PhantomData<&'s ()>`. I think `PhantomData<&'s DS>` is a little misleading - because it's not pretending to encompass a reference to `DS`, but actually to some underlying store.